<!doctype HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html lang="EN">
<!-- ABENCLASS_EXCEPTION_GUIDL -->

<!-- Mirrored from help.sap.com/doc/abapdocu_754_index_htm/7.54/en-US/abenclass_exception_guidl.htm by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 18 Feb 2024 14:53:06 GMT -->
<head>
<meta http-equiv="content-type" content="text/html; charset=utf-8">
<meta name="author" content="SAP">
<meta name="description" content="Classic and Class-Based Exceptions">
<title>
Classic and Class-Based Exceptions - ABAP Keyword Documentation
</title>
<link rel="icon" href="ABAPIcon.ico" type="image/ico">
<link rel="shortcut icon" href="ABAPIcon.ico">
<LINK rel="stylesheet" href="abap_docu.css" type="text/css">
<script language="javascript" type="text/javascript" src="functions.js"></script>
<script type="text/JavaScript">
function urlToClipboard(id){
 try{ var loc = top.location.href;}
 catch(e){ alert( "URL cannnot be determined" ); return; }
 var field = document.getElementById(id);
 field.focus();
 if ( loc.search( ".htm" ) == -1 && loc.search( ".HTM" ) == -1  ){ loc = loc + "index.htm"; }
 if (   loc.search( "index.html" ) != -1
     || loc.search( "INDEX-2.html" ) != -1 ){
   var off = loc.search( "file=" );
   if ( off == -1  ){
     field.value = loc + "?file=abenclass_exception_guidl.htm"; }
   else {
     field.value = loc.substring( 0, off ) + "file=abenclass_exception_guidl.htm";
   }
 }
 else {
    field.value = loc; }
 field.setSelectionRange(0, field.value.length);
 var r = confirm( "Copy URL to clipboard?\n\n" + field.value + "\n\nCopy to clipboard does not work in all browsers." );
 if (r == true) {
   try{
     document.execCommand("copy");
   }
     catch(e){ alert( "URL could not be copied to clipboard" ); }
 }
 window.scrollTo(0,0);
}
</script>
</head>
<body>
<div class="all">
<div style="float:right;">
<input type="button" value="URL"
style="font-size:10px; width:30px; height:18px; padding:0; "
onclick="urlToClipboard('url')" title="Copy URL to Clipboard" ></div>
<p>
<input name="query" id="query" value="" type="text" title="Suchbegriff" size="30" maxlength="30" onkeyup="getInput()">&nbsp;
<input type="button" value="Search"
style="font-size:12px; width:50px; height:18px; padding:0"
onclick="call_search('search.html')" title="Search in ABAP Keyword Documentation" >
<hr style="width:100%;height:1px"></p>
<p class="copyright">AS ABAP Release 754, &copy;Copyright 2019 SAP SE. All rights reserved.</p>
<span class="path">
<a href="javascript:call_link('abenabap.html')" class="blue">ABAP Keyword Documentation</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenabap_pgl.html')" class="blue">ABAP Programming Guidelines</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenarchitecture_guidl.html')" class="blue">Architecture</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenerror_handling_guidl.html')" class="blue">Error Handling</a>&nbsp;&rarr;&nbsp;
</span>
<p><span class="h1">
Classic and Class-Based Exceptions
</span></p>
<A name="@@ITOC@@ABENCLASS_EXCEPTION_GUIDL_1"></A>
<p><span class="h2">
Background
</span></p>
<p>
For reasons of downward compatibility, there are two options for defining standalone handleable exceptions in ABAP:
</p>
<ul class="disc">
<li><span class="bold">Classic Exceptions</span> <br>These exceptions can only be declared in the interfaces
of methods or function modules using <span class="qtext">EXCEPTIONS</span> and can be raised within
such a procedure using the statements <a href="javascript:call_link('abapraise_exception.html')" class="blue"><span class="qtext">RAISE</span></a>
or <a href="javascript:call_link('abapmessage_raising.html')" class="blue"><span class="qtext">MESSAGE RAISING</span></a>. The
procedure caller can use the addition <span class="qtext">EXCEPTIONS</span> of the statements <a href="javascript:call_link('abapcall_method_static_short.html')" class="blue"><span class="qtext">meth( ...
)</span></a> or <a href="javascript:call_link('abapcall_function.html')" class="blue"><span class="qtext">CALL FUNCTION</span></a>
to assign return codes for the system field <span class="qtext">sy-subrc</span> to the exceptions the caller wants to handle and evaluate them after the call.</li>
</ul>
<ul class="disc">
<li><span class="bold">Class-Based Exceptions</span> <br>These exceptions are defined by exception classes,
from which an exception object can be created when an exception is raised (if a handler uses the addition
<span class="qtext">INTO</span> in <span class="qtext">CATCH</span>). A class-based exception can either
cancel the current context or allow for a resume. Exceptions are raised using the statement <span class="qtext">RAISE
EXCEPTION</span> and handled using <span class="qtext">CATCH</span> in a <span class="qtext">TRY</span>
control structure. Class-based exceptions can be raised in any procedures and can be further propagated by any procedures.</li>
</ul>
<p>
The coexistence of the two exception concepts is regulated as follows:
</p>
<ul class="disc">
<li>Classic and class-based exceptions cannot be declared together in the interface of a procedure. Within a processing block, either only classic or only class-based exceptions can be raised.</li>
</ul>
<ul class="disc">
<li>For reasons of interoperability, within a processing block class-based exceptions can be handled and evaluate the return values of function modules and methods using classic exceptions.</li>
</ul>
<A name="@@ITOC@@ABENCLASS_EXCEPTION_GUIDL_2"></A>
<p><span class="h2">
Rule
</span></p>
<p>
<span class="bold">Using Class-Based Exceptions</span>
</p>
<p>
Only raise class-based exceptions in new procedures if it is possible to dispense with classic exceptions from the technical point of view.
</p>
<A name="@@ITOC@@ABENCLASS_EXCEPTION_GUIDL_3"></A>
<p><span class="h2">
Details
</span></p>
<p>
Self-defined classic exceptions are little more than return values. If a classic exception is raised
in a procedure using the statement <span class="qtext">RAISE</span>, the <span class="qtext">sy-subrc</span>
system field is set according to the raised exception after the return to the calling program. The calling
program itself must always check, by querying <span class="qtext">sy-subrc</span>, whether an exception
was raised and react to it if required, for example, by appropriate handling or explicit forwarding
to its own calling program (by raising a separate equivalent exception). This does not improve the clarity of the program.
</p>
<p>
The raising of class-based exceptions, however, results in a change of the program flow. They can either be handled directly or propagated upwards along the call hierarchy. In this way, not every procedure
(<a href="javascript:call_link('abenfunct_module_subroutine_guidl.html')" class="lnkpurple" title="Guideline">method</a>) has to consider every possible exception situation itself. This supports the
<a href="javascript:call_link('abenseperation_concerns_guidl.html')" class="lnkpurple" title="Guideline">separation of concerns</a>
within an application. Because the exception can be represented by an object of an exception class,
this exception object can gather additional information about the exception situation and transport it to the handler. In contrast to classic exceptions, this can also include specific
<a href="javascript:call_link('abenexception_texts_guidl.html')" class="lnkpurple" title="Guideline">exception texts</a>.
</p>
<p>
By default, raising an exception stops the entire current context even if the exception is handled.
However, there may be situations (mass data processing, for instance) in which a single error does not justify canceling an entire service. For these cases, class-based exceptions can be raised and propagated as resumable exceptions
(<span class="qtext">RESUMABLE</span>). A handler can decide whether a service is canceled completely
or is resumed using the statement <span class="qtext">RESUME</span>, for example after a corresponding log entry has been written.
</p>
<p>
Class-based exceptions completely replace the classic exceptions for new code (of course, there are
exceptions to this rule) and add resumability. Although classic exceptions on the raiser side are completely
obsolete from a technical point of view, the following must be considered for older code: Even if the
raiser side is under control, it is not simply a case of switching older procedures over to class-based exceptions, because then all usage occurrences would have to be modified.
</p>
<p>
When existing procedures that use classic exceptions are called, they must continue to be handled in
the new code. In this case, we recommend mapping the classic exceptions to equivalent class-based exceptions
by using <span class="qtext">RAISE EXCEPTION</span>. In this way, class-based error handling is provided
that is uniform to the outside. The exception situation can then be forwarded to higher call layers without each layer having to react to this situation explicitly.
</p>
<p><span class="h4">
Exception
</span></p>
<p>
Since class-based exceptions are currently not supported in remote-enabled function modules (RFM), classic exceptions still need to be implemented and handled for remote function calls (RFCs).
</p>
<p><span class="h4">
Bad Example
</span></p>
<p>
The following source code shows the declaration and the raising of a classic exception in a method as
well as their handling by evaluating <span class="qtext">sy-subrc</span> after a call of the method. This procedure infringes the above rule. .
</p>
<dl><dd>
<span class="qtext">CLASS application DEFINITION. <br>&nbsp;&nbsp;PUBLIC SECTION. <br>&nbsp;&nbsp;&nbsp;&nbsp;METHODS do_something <br>&nbsp;&nbsp;&nbsp;&nbsp;EXCEPTIONS application_error. <br>ENDCLASS.</span>
</dd></dl>
<dl><dd>
<span class="qtext">CLASS application IMPLEMENTATION. <br>&nbsp;&nbsp;METHOD do_something. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>
&nbsp;&nbsp;&nbsp;&nbsp;RAISE application_error. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>&nbsp;&nbsp;ENDMETHOD. <br>ENDCLASS.</span>
</dd></dl>
<dl><dd>
<span class="qtext">...</span>
</dd></dl>
<dl><dd>
<span class="qtext">... oref TYPE REF TO application. <br> <br>... <br> <br>oref->do_something( <br>&nbsp;&nbsp;EXCEPTIONS application_error = 4 ). <br>IF sy-subrc &lt;> 0. <br>&nbsp;&nbsp;... <br>ENDIF.</span>
</dd></dl>
<p><span class="h4">
Good Example
</span></p>
<p>
The following source code shows the definition of an exception class, its declaration, and the raising
in a method as well as its handling using <span class="qtext">CATCH</span> after the call of the method in a <span class="qtext">TRY</span> block.
</p>
<dl><dd>
<span class="qtext">CLASS cx_application_error DEFINITION <br>&nbsp;&nbsp;INHERITING FROM cx_static_check. <br>ENDCLASS.</span>
</dd></dl>
<dl><dd>
<span class="qtext">CLASS application DEFINITION. <br>&nbsp;&nbsp;PUBLIC SECTION. <br>&nbsp;&nbsp;&nbsp;&nbsp;METHODS do_something <br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;RAISING cx_application_error. <br>ENDCLASS.</span>
</dd></dl>
<dl><dd>
<span class="qtext">CLASS application IMPLEMENTATION. <br>&nbsp;&nbsp;METHOD do_something. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>
&nbsp;&nbsp;&nbsp;&nbsp;RAISE EXCEPTION TYPE cx_application_error. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>&nbsp;&nbsp;ENDMETHOD. <br>ENDCLASS.</span>
</dd></dl>
<dl><dd>
<span class="qtext">...</span>
</dd></dl>
<dl><dd>
<span class="qtext">... oref TYPE REF TO application. <br> <br>... <br> <br>TRY. <br>&nbsp;&nbsp;&nbsp;&nbsp;oref->do_something( ). <br>
&nbsp;&nbsp;&nbsp;&nbsp;CATCH cx_application_error. <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;... <br>ENDTRY.</span>
</dd></dl>
<p>
This simple example is perhaps not the most obvious demonstration of the great advantage of class-based
exceptions over classic exceptions. However, the advantage is clearly seen in nested procedure calls and the handling of exceptions that were raised in more distant call levels.
</p>
<br>
<br>
<br>
<br>
</DIV>
<fieldset style="opacity:0">
<input id="url" type="text" />
</fieldset>
</BODY>

<!-- Mirrored from help.sap.com/doc/abapdocu_754_index_htm/7.54/en-US/abenclass_exception_guidl.htm by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 18 Feb 2024 14:53:06 GMT -->
</html>
